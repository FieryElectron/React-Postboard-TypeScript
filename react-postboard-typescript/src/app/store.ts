import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';

import signInPageReducer from '../components/SignInPage/signInPageSlice';
import postBoardPageReducer from '../components/PostBoardPage/postBoardPageSlice';

export const store = configureStore({
  reducer: {
    signInPage: signInPageReducer,
    postBoardPage: postBoardPageReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
