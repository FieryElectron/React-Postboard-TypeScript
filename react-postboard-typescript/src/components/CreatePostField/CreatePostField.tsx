import styles from './CreatePostField.module.css';

import Button from '../Button/Button';
import Input from '../Input/Input';
import { toast } from 'react-toastify';

import axios from 'axios';

import { useNavigate } from 'react-router-dom';
import { useAppSelector, useAppDispatch } from '../../app/hooks';

import {
    selectPostContent,
    selectPost,
    selectToken,
    getPosts,
    setPostContent,
} from '../PostBoardPage/postBoardPageSlice';

const refreshAccessToken = async (tokenUrl: string) => {
    const res = await axios.get(tokenUrl, {withCredentials: true});
    return res.data.flag;
}

const CreatePostField = () => {
    const dispatch = useAppDispatch();
    const navigate = useNavigate();

    const postContent = useAppSelector(selectPostContent);
    const postUrl = useAppSelector(selectPost);
    const tokenUrl = useAppSelector(selectToken);


    const tryCreatePost = async () => {
        const res = await axios.post(postUrl, {content:postContent}, { withCredentials: true });

        if (res.data.flag) {
            toast.info(res.data.info, {
                position: 'top-center',
                autoClose: 1000,
                pauseOnFocusLoss:false,
            });

            dispatch(getPosts(navigate));
            return;
        }

        if(res.data.info){
            toast.info(res.data.info, {
                position: 'top-center',
                autoClose: 1000,
                pauseOnFocusLoss:false,
            });
        }


        if (res.data.refreshToken) {
            if (await refreshAccessToken(tokenUrl)) {
                const resSecond = await axios.post(postUrl, {content:postContent}, { withCredentials: true });
                if (resSecond.data.flag) {
                    toast.info(res.data.info, {
                        position: 'top-center',
                        autoClose: 1000,
                        pauseOnFocusLoss:false,
                    });
        
                    dispatch(getPosts(navigate));
                    return;
                }
            }
        }
    
        // navigate('/');
        return;
    }

    const contentOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        dispatch(setPostContent(e.target.value));
    }

    return (
        <div data-testid='createPostFieldContainer'  className={styles.createPostFieldContainer}>
            <Input placeholder='Post Content' onChange={contentOnChange}/>
            <Button text='Create' color='#38DD38' onClick={tryCreatePost} />
        </div>
    )
}

CreatePostField.defaultProps = {
    text: 'text',
}

export default CreatePostField;