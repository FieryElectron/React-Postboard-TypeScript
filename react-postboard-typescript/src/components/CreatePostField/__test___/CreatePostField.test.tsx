import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import '@testing-library/jest-dom/extend-expect';

import CreatePostField from '../CreatePostField';
import { Provider } from 'react-redux';

import { BrowserRouter as Router } from 'react-router-dom';
import { store } from '../../../app/store';

let container: HTMLElement;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container.remove();
});

it('empty test', () => {

});

it('renders CreatePostField without crashing', () => {
  act(() => {
    render(<Provider store={store}><Router><CreatePostField /></Router></Provider>);
  });
});

it('renders CreatePostField class correctly', () => {
  act(() => {
    render(<Provider store={store}><Router><CreatePostField /></Router></Provider>);
  });

  const createPostFieldContainer = document.querySelector('[data-testid=createPostFieldContainer]')!;
  expect(createPostFieldContainer).toHaveAttribute('class', 'createPostFieldContainer');
});

it('matchs snapshot 1', ()=>{
    const tree = renderer.create(<Provider store={store}><Router><CreatePostField/></Router></Provider>).toJSON();
    expect(tree).toMatchSnapshot();
}); 




