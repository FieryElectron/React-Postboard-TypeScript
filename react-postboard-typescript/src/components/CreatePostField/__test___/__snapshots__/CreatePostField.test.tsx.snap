// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchs snapshot 1 1`] = `
<div
  className="createPostFieldContainer"
  data-testid="createPostFieldContainer"
>
  <div
    className="inputContainer"
    data-testid="inputContainer"
  >
    <input
      className="input"
      data-testid="input"
      onChange={[Function]}
      placeholder="Post Content"
      spellCheck="false"
      type="text"
    />
  </div>
  <button
    className="button"
    data-testid="button"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#38DD38",
      }
    }
  >
    Create
  </button>
</div>
`;
